src/pybind/mgr/volumes/fs/volume.py

```diff
@@ -163,16 +303,49 @@ def list_volumes(self):
             result.append({'name': f['mdsmap']['fs_name']})
         return 0, json.dumps(result, indent=2), ""
 
+    def connection_pool_wrap(func):
+        """
+        decorator that wraps subvolume calls by fetching filesystem handle
+        from the connection pool when fs_handle argument is empty, otherwise
+        just invoke func with the passed in filesystem handle. Also handles
+        call made to non-existent volumes (only when fs_handle is empty).
+        """
+        def conn_wrapper(self, fs_handle, **kwargs):
+            fs_h = fs_handle
+            fs_name = kwargs['vol_name']
+            # note that force arg is available for remove type commands
+            force = kwargs.get('force', False)
+
+            # fetch the connection from the pool
+            if not fs_handle:
+                try:
+                    fs_h = self.connection_pool.get_fs_handle(fs_name)
+                except VolumeException as ve:
+                    if not force:
+                        return self.volume_exception_to_retval(ve)
+                    return 0, "", ""
+
+            # invoke the actual routine w/ fs handle
+            result = func(self, fs_h, **kwargs)
+
+            # hand over the connection back to the pool
+            if not fs_handle:
```

Comment:
If there is no fs_handle, then you put back the fs handle in the connection pool? I'm not following here.

Reply:
`fs_handle` is what's passed to this decorator (the decorator does not mangle that) -- when empty, the filesystem handle is fetched from the connection pool and has to be "put" back into the pool. Probably, rewriting this condition as

```
if fs_h:
    self.connection_pool.put_fs_handle(fs_name)
```

makes it more clear?

Reply:
Yes. That's what I expected.

Reply:
true -- I just tried to mimic the initial condition (above) to fetch the file handle from pool...